SSL Certificate and CORS

When setting up Flynn, you're being asked to download and install the certificate. IIRC this only works on per-browser basis. Whenever another client (different browser instance) logs into the dashboard, he'll see a warning:

![flynn1](https://cloud.githubusercontent.com/assets/3372410/7004132/c4ee5906-dc66-11e4-952b-94790e5a3252.PNG)

Once added, he can sign in and proceed to the apps tab, but he won't see any:

![flynn2](https://cloud.githubusercontent.com/assets/3372410/7004139/d5352704-dc66-11e4-90f1-e16d146ef9ae.PNG)

This is because the security exception applies only for this specific domain.

I suggest to add a warning to the dashboard which instructs the user to download the certificate if network calls are blocked due to insecure responses.

If anyone can hint me at the proper direction for this, I'd  be happy to create a PR.

I'm not able to run make:

```
`[ tup ] [37.694s] Executing Commands...
 [                                               ETA~=??? Remaining=126 Active=2                                                ]   0%.tup/tmp/output-17974: Text file busy
tup error: Unable to unlink sub-process output file.
 *** Command ID=17974 failed: ./build.sh
 [                                               ETA~=??? Remaining=126 Active=1                                                ]   0%.tup/tmp/output-17984: Text file busy
tup error: Unable to unlink sub-process output file.
 *** Command ID=17984 failed: CGO_ENABLED=0 ../../util/_toolchain/go/bin/go build -installsuffix nocgo -o bin/flynn-postgres -ldflags="-X github.com/flynn/flynn/pkg/version.commit $GIT_COMMIT -X github.com/flynn/flynn/pkg/version.branch $GIT_BRANCH -X github.co
m/flynn/flynn/pkg/version.tag $GIT_TAG -X github.com/flynn/flynn/pkg/version.dirty $GIT_DIRTY" -tags="$GO_BUILD_TAGS"
 [                                               ETA~=??? Remaining=126 Active=0                                                ]   0%
 *** tup: 2 jobs failed.
make: *** [all] Error 1

```

@current master

--- a/dashboard/app/lib/javascripts/dashboard/client.js
+++ b/dashboard/app/lib/javascripts/dashboard/client.js
@@ -97,6 +97,15 @@ Dashboard.Client = Marbles.Utils.createClass({
 		});
 	},
 
+	ping: function (endpoint, protocol) {
+		endpoint = window.location.host.replace("dashboard", endpoint);
+		endpoint = protocol + "://" + endpoint + "/ping";
+
+		return this.performRequest('GET', {
+			url: endpoint
+		});
+	},
+
 	getApps: function () {
 		return this.performControllerRequest('GET', {
 			url: "/apps"

--- a/dashboard/app/lib/javascripts/dashboard/config.js
+++ b/dashboard/app/lib/javascripts/dashboard/config.js
@@ -88,7 +88,7 @@ Dashboard.config.fetch = function () {
 
 	Marbles.HTTP({
 		method: 'GET',
-		url: Dashboard.config.API_SERVER + "/config",
+		url: Dashboard.config.API_SERVER.replace(/^https?:/, window.location.protocol) + "/config",
 		middleware: [
 			Marbles.HTTP.Middleware.WithCredentials,
 			Marbles.HTTP.Middleware.SerializeJSON

--- a/dashboard/app/lib/javascripts/dashboard/main.js
+++ b/dashboard/app/lib/javascripts/dashboard/main.js
@@ -13,8 +13,20 @@ window.Dashboard = {
 	config: {},
 
 	waitForRouteHandler: Promise.resolve(),
+	errCertNotInstalled: new Error("HTTPS certificate is not trusted."),
+	errServiceUnavailable: new Error("Service is unavailable."),
 
 	run: function () {
+		var self = this, loadURL, resolveWaitForNav;
+		this.waitForNav = new Promise(function(resolve, reject) {
+			resolveWaitForNav = resolve;
+		});
+
+		loadURL = function() {
+			resolveWaitForNav();
+			Marbles.history.loadURL();
+		};
+
 		if ( Marbles.history && Marbles.history.started ) {
 			throw new Error("Marbles.history already started!");
 		}
@@ -38,11 +50,11 @@ window.Dashboard = {
 			dispatcher: this.Dispatcher,
 			trigger: false
 		});
-		if (window.location.protocol === "http:" && this.config.INSTALL_CERT) {
-			Marbles.history.navigate("installcert");
-			Marbles.history.loadURL();
+
+		if (this.config.INSTALL_CERT) {
+			this.__isCertInstalled().then(loadURL);
 		} else {
-			Marbles.history.loadURL();
+			loadURL();
 		}
 	},
 
@@ -63,6 +75,92 @@ window.Dashboard = {
 	__redirectToLogin: function () {
 		var redirectPath = Marbles.history.path ? '?redirect='+ encodeURIComponent(Marbles.history.path) : '';
 		Marbles.history.navigate('login'+ redirectPath);
+	},
+
+	__catchInsecurePingResponse: function(httpsArgs) {
+		var httpsXhr = httpsArgs[1], self = this,
+			handleSuccess, handleError;
+
+		handleSuccess = function (httpArgs) {
+			var httpXhr = httpArgs[1];
+			// https did not work but http did...something is wrong with the cert
+			self.Dispatcher.handleAppEvent({
+				name: "HTTPS_CERT_MISSING",
+				status: httpXhr.status
+			});
+			return Promise.reject(self.errCertNotInstalled);
+		};
+		handleError = function (httpArgs) {
+			var httpXhr = httpArgs[1];
+			if (httpArgs.length === 1) {
+				return httpArgs;
+			}
+
+			if (httpXhr.status === 0) {
+				// https is failing as well...service is unavailable
+				self.Dispatcher.handleAppEvent({
+					name: "SERVICE_UNAVAILABLE",
+					status: httpXhr.status
+				});
+				return Promise.reject(self.errServiceUnavailable);
+			}
+			// https did not work but http did without a network error
+			// => missing ssl exception for controller
+			self.Dispatcher.handleAppEvent({
+				name: "HTTPS_CERT_MISSING",
+				status: httpXhr.status
+			});
+			return Promise.reject(self.errCertNotInstalled);
+		};
+
+		if (httpsXhr.status === 0) {
+			// https is unavailable, let's see if http works
+			return self.client.ping("controller", "http").then(handleSuccess).catch(handleError);
+		}
+		// an error code other than 0
+		self.Dispatcher.handleAppEvent({
+			name: "SERVICE_UNAVAILABLE",
+			status: httpsXhr.status
+		});
+		return Promise.reject(this.errServiceUnavailable);
+	},
+
+	__catchSecurePingResponse: function(args) {
+		var xhr = args[1];
+		if (xhr.status === 0) {
+			// We were not able to access the controller due to a network error (ssl, timeout)
+			// In order to understand what's happening, we have to switch to http.
+			this.Dispatcher.handleAppEvent({
+				name: "CONTROLLER_UNREACHABLE_FROM_HTTPS",
+				status: xhr.status
+			});
+			return Promise.reject(new Error("CONTROLLER_UNREACHABLE_FROM_HTTPS"));
+		}
+
+		// an error code other than 0
+		this.Dispatcher.handleAppEvent({
+			name: "SERVICE_UNAVAILABLE",
+			status: xhr.status
+		});
+		return Promise.reject(this.errServiceUnavailable);
+	},
+
+	__successPingResponse: function(args) {
+		var xhr = args[1];
+		if (xhr.status === 200) {
+			window.location.href = window.location.href.replace("http:", "https:");
+		}
+	},
+
+	__isCertInstalled: function() {
+		var self = this;
+		if (window.location.protocol === "https:") {
+			return this.client.ping("controller", "https").catch(this.__catchSecurePingResponse.bind(this));
+		} else {
+			return this.client.ping("controller", "https")
+				.then(this.__successPingResponse.bind(this))
+				.catch(this.__catchInsecurePingResponse.bind(this));
+		}
 	},
 
 	__handleEvent: function (event) {
@@ -95,11 +193,7 @@ window.Dashboard = {
 	__handleAppEvent: function (event) {
 		switch (event.name) {
 			case "CONFIG_READY":
-				if (window.location.protocol === "http:" && this.config.INSTALL_CERT) {
-					window.location.href = window.location.href.replace("http", "https");
-				} else {
-					this.__handleConfigReady();
-				}
+				this.__handleConfigReady();
 			break;
 
 			case "AUTH_CHANGE":
@@ -110,12 +204,17 @@ window.Dashboard = {
 				this.__handleGithubAuthChange(event.authenticated);
 			break;
 
+			case "CONTROLLER_UNREACHABLE_FROM_HTTPS":
+				// Controller isn't accessible via https. Redirect to http and try again.
+				window.location.href = window.location.href.replace("https", "http");
+			break;
+
+			case "HTTPS_CERT_MISSING":
+				Marbles.history.navigate("installcert", {force: true});
+			break;
+
 			case "SERVICE_UNAVAILABLE":
-				if (window.location.protocol === "http:" && this.config.INSTALL_CERT) {
-					this.__handleConfigReady();
-				} else {
-					this.__handleServiceUnavailable(event.status);
-				}
+				this.__handleServiceUnavailable(event.status);
 			break;
 		}
 	},
@@ -132,12 +231,16 @@ window.Dashboard = {
 		this.__renderNavComponent();
 
 		if ( !authenticated && !this.__isLoginPath() ) {
-			var currentHandler = Marbles.history.getHandler();
+			var currentHandler = Marbles.history.getHandler(),
+				self = this;
 			if (currentHandler && currentHandler.opts.auth === false) {
 				// Don't redirect to login from page not requiring auth
 				return;
 			}
-			this.__redirectToLogin();
+
+			this.waitForNav.then(function() {
+				self.__redirectToLogin();
+			});
 		}
 	},
 

